utils/pattern/src/single.rs

```diff
@@ -0,0 +1,256 @@
+/// after that initial code point. If zero, there is no placeholder.
+///
+/// # Examples
+///
+/// Parsing a pattern into the encoding:
+///
+/// ```
+/// use icu_pattern::Pattern;
+/// use icu_pattern::SinglePlaceholder;
+///
+/// // Parse the string syntax and check the resulting data store:
+/// let store = Pattern::<SinglePlaceholder, _>::try_from_str("Hello, {0}!")
+///     .unwrap()
+///     .take_store();
+///
+/// assert_eq!("\u{8}Hello, !", store);
+/// ```
+///
+/// Example patterns supported by this backend:
+///
+/// ```
+/// use icu_pattern::Pattern;
+/// use icu_pattern::SinglePlaceholder;
+///
+/// // Single numeric placeholder:
+/// assert_eq!(
+///     Pattern::<SinglePlaceholder, _>::try_from_str("{0} days ago")
+///         .unwrap()
+///         .interpolate_to_string([5]),
+///     "5 days ago",
+/// );
+///
+/// // Single named placeholder:
+/// assert_eq!(
+///     Pattern::<SinglePlaceholder, _>::try_from_str("{name}")
+///         .unwrap()
+///         .interpolate_to_string(["Alice"]),
+///     "Alice",
+/// );
+///
+/// // No placeholder (note, the placeholder value is never accessed):
+/// assert_eq!(
+///     Pattern::<SinglePlaceholder, _>::try_from_str("yesterday")
+///         .unwrap()
+///         .interpolate_to_string(["hi"]),
+///     "yesterday",
+/// );
+///
+/// // Escaped placeholder and a real placeholder:
+/// assert_eq!(
+///     Pattern::<SinglePlaceholder, _>::try_from_str("'{0}' {1}")
+///         .unwrap()
+///         .interpolate_to_string(("hi",)),
+///     "{0} hi",
+/// );
+/// ```
+///
+/// [`Pattern::interpolate()`]: crate::Pattern::interpolate
+#[derive(Debug)]
+pub struct SinglePlaceholder {
+    _not_constructible: core::convert::Infallible,
+}
+
+impl crate::private::Sealed for SinglePlaceholder {}
+
+impl PatternBackend for SinglePlaceholder {
+    type PlaceholderKey = SinglePlaceholderKey;
+    type Store = str;
+    type Iter<'a> = SinglePlaceholderPatternIterator<'a>;
+
+    fn validate_store(store: &Self::Store) -> Result<(), Error> {
+        let placeholder_offset_char = store.chars().next().ok_or(Error::InvalidPattern)?;
+        let initial_offset = char::len_utf8(placeholder_offset_char);
+        let placeholder_offset = placeholder_offset_char as usize;
+        if placeholder_offset > store.len() - initial_offset + 1 {
+            return Err(Error::InvalidPattern);
+        }
+        Ok(())
+    }
+
+    fn iter_items(store: &Self::Store) -> Self::Iter<'_> {
+        let placeholder_offset_char = match store.chars().next() {
+            Some(i) => i,
+            None => {
+                debug_assert!(false);
+                '\0'
+            }
+        };
+        let initial_offset = char::len_utf8(placeholder_offset_char);
+        SinglePlaceholderPatternIterator {
+            store,
+            placeholder_offset: placeholder_offset_char as usize + initial_offset - 1,
+            current_offset: initial_offset,
+        }
```

Comment:
I think it's fine to precompute the three values, saves you quite some logic.

```suggestion
        let split = placeholder_offset_char as usize + initial_offset;
        [PatternItem::String(&store[..split]), PatternItem::Placeholder(..), PatternItem::String(&store[split..])].into_iter()
```

In fact, it would be nice if I could obtain the array instead of just an iterator.

Reply:
Ok, but the three values don't always exist. I don't return empty strings as elements, and the placeholder is optional. The logic I wrote is intended to work for all the different cases, most of which have at least a docs test.